### PR TITLE
refactor: global config

### PR DESCRIPTION
**Pull request by pret-a-porter**
## Changes:

Refactored global config

## Context:

#12639

## Documentation

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/config/global.ts [hunk 1]
```diff
@@ -1,40 +1,56 @@
...
+
+  static get(): RepoGlobalConfig;
+  static get<Key extends keyof RepoGlobalConfig>(
+    key?: Key
+  ): RepoGlobalConfig[Key];
+  static get<Key extends keyof RepoGlobalConfig>(
+    key?: Key
+  ): RepoGlobalConfig | RepoGlobalConfig[Key] {
+    return key ? GlobalConfig.config[key] : GlobalConfig.config;
   }
```
Comment (JamieMagee) [inline lib/config/global.ts:35]: Interesting! You can define the overloads, but you don't have to explicitly call them. I've not seen this pattern before :smile:

File: lib/config/global.ts [hunk 2]
```diff
@@ -1,40 +1,56 @@
...
-  config: RenovateConfig | RepoGlobalConfig = {}
-): RenovateConfig {
-  repoGlobalConfig = {};
-  const result = { ...config };
-  for (const option of repoGlobalOptions) {
-    repoGlobalConfig[option] = config[option];
-    delete result[option];
+export class GlobalConfig {
+  // TODO: once global config work is complete, add a test to make sure this list includes all options with globalOnly=true (#9603)
+  private static readonly OPTIONS = [
```
Comment (JamieMagee) [inline lib/config/global.ts:5]: I tried making this `(keyof RepoGlobalConfig)[]`, but had issues in `set` where `config[option]` was `unknown`.

It's probably not possible to do, but if you've got any ideas?

**Comment by JamieMagee** [comment]
Can you run on a real repo please? To confirm end-to-end flow. Apart from that, this looks good to go!

**Comment by pret-a-porter** [comment]
> Can you run on a real repo please? To confirm end-to-end flow. Apart from that, this looks good to go!

Sure. Shall I configure it some certain way?

File: lib/config/global.ts [hunk 3]
```diff
@@ -1,40 +1,56 @@
...
+    return result;
+  }
+
+  static reset(): void {
+    GlobalConfig.config = {};
+  }
+
+  static get isDryRun(): boolean {
+    return GlobalConfig.get('dryRun');
+  }
```
Comment (viceice) [inline lib/config/global.ts:55]: Do we really need this?

File: lib/config/global.ts (hunk 1 above)
Comment (viceice) [reply lib/config/global.ts:35]: Yeah, typescript will only use first two decalrations for caller type check and the last for local type check

File: lib/config/global.ts (hunk 3 above)
Comment (pret-a-porter) [reply lib/config/global.ts:55]: Not really, It is optional thing

**Comment by pret-a-porter** [comment]
> Yeah, typescript will only use first two decalrations for caller type check and the last for local type check

Done https://github.com/pret-a-porter-testing/renovate-reproduction-12639

File: lib/config/global.ts (hunk 3 above)
Comment (viceice) [reply lib/config/global.ts:55]: i don't think we need this, as `GlobalConfig.get('dryRun')` isn't much longer than `GlobalConfig.isDryRun`. So i prefer to remove this

**Review by viceice — COMMENTED** [review]
Needs some more changes because of main merge 😉

**Review by viceice — APPROVED** [review]

**Pull request merged**